src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py

```diff
@@ -0,0 +1,194 @@
+        wmaSpecs = self.getActiveWorkflows()
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow specs from upstream ReqMgr2 server, the returned dict
+        # is composed by workflow names and JSON dict representing the workflow
+        specs = self.getRequestSpecs(wflows)
+
+        # iterate over workflow items and update local WorkQueue and pickle files if
+        # either site white or black lists are different
+        for wflow, rdict in specs.items():
+            siteWhiteList = rdict['SiteWhitelist']
+            siteBlackList = rdict['SiteBlacklist']
+
+            # get the name of pkl file from wma spec
+            pklFileName = wmaSpecs[wflow]
+
+            # create wrapper helper and load pickle file
+            wHelper = WMWorkloadHelper()
+            wHelper.load(pklFileName)
+
+            # extract from pickle spec both white and black site lists and compare them
+            # to one we received from upstream service (ReqMgr2)
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if wmaWhiteList != siteWhiteList or wmaBlackList != siteBlackList:
+                self.logger.info(f"Updating {wflow}: siteWhiteList {wmaWhiteList} => {siteWhiteList} and siteBlackList {wmaBlackList} => {siteBlackList}")
+                try:
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
+                    logging.exception(msg)
+                    raise LocalWorkqueueUpdateException(msg) from None
+
+                # update workload only if we updated local WorkQueue
+                # update site white/black lists together
+                wHelper.setWhitelist(siteWhiteList)
+                wHelper.setBlacklist(siteBlackList)
+
+                try:
+                    # persist the spec in local CouchDB
+                    self.logger.info(f"Updating {self.localCouchUrl} with new site lists for {wflow}")
+                    wHelper.saveCouchUrl(self.localCouchUrl)
+
+                    # save back pickle file
+                    newPklFileName = pklFileName.split('.pkl')[0] + '_new.pkl'
+                    wHelper.save(newPklFileName)
+
+                    # if new pickle file is saved we can swap it with original one
+                    if os.path.getsize(newPklFileName) > 0:
+                        self.logger.info(f"Updated {pklFileName}")
+                        shutil.move(newPklFileName, pklFileName)
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
+                    logging.exception(msg)
+                    raise SpecFileUpdateException(msg) from None
```

Comment:
Same as the comment made for the other exception. We cannot break the whole cycle, instead, the component needs to keep processing further workflows and simply skip the one where we had problems updating the list.

Reply:
ok, switched to continue